metrics-core/src/lib.rs

```diff
@@ -32,24 +32,172 @@
-/// callers, though, we opt to use a copy-on-write pointer -- [`Cow`] -- to allow callers
-/// flexiblity in how and what they pass.
-pub type Key = Cow<'static, str>;
+/// We specify `ScopedString` to attempt to get the best of both worlds: flexibility to provide a
+/// static or dynamic (owned) string, while retaining the performance benefits of being able to
+/// take ownership of owned strings and borrows of completely static strings.
+pub type ScopedString = Cow<'static, str>;
 
-/// A value which can be converted into a nanosecond representation.
+/// A key/value pair used to further describe a metric.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Label(ScopedString, ScopedString);
+
+impl Label {
+    /// Creates a `Label` from a key and value.
+    pub fn from_parts<K, V>(key: K, value: V) -> Self
+    where
+        K: Into<ScopedString>,
+        V: Into<ScopedString>,
+    {
+        Label(key.into(), value.into())
+    }
+}
+
+/// A metric key.
+///
+/// A key always includes a name, but can optional include multiple labels used to further describe
+/// the metric.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Key {
+    #[doc(hidden)]
+    pub name: ScopedString,
+    #[doc(hidden)]
+    pub labels: Option<Vec<Label>>,
+}
+
+impl Key {
+    /// Creates a `Key` from a name.
+    pub fn from_name<N>(name: N) -> Self
+    where
+        N: Into<ScopedString>,
+    {
+        Key {
+            name: name.into(),
+            labels: None,
+        }
+    }
+
+    /// Creates a `Key` from a name and vector of `Label`s.
+    pub fn from_name_and_labels<N, L>(name: N, labels: L) -> Self
+    where
+        N: Into<ScopedString>,
+        L: Into<Vec<Label>>,
+    {
+        Key {
+            name: name.into(),
+            labels: Some(labels.into()),
+        }
+    }
+
+    /// Name of this key.
+    pub fn name(&self) -> ScopedString {
+        self.name.clone()
+    }
+
+    /// Labels of this key, if they exist.
+    pub fn labels(&self) -> Option<Iter<Label>> {
+        self.labels.as_ref().map(|i| i.iter())
+    }
+
+    /// Maps the name of this `Key` to a new name.
+    pub fn map_name<F>(self, f: F) -> Self
+    where
+        F: FnOnce(ScopedString) -> ScopedString,
+    {
+        Key {
+            name: f(self.name),
+            labels: self.labels,
+        }
+    }
+}
+
+impl fmt::Display for Key {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self.labels {
+            None => write!(f, "Key({}", self.name),
+            Some(labels) => {
+                let kv_pairs = labels
+                    .iter()
+                    .map(|label| format!("{} = {}", label.0, label.1))
+                    .collect::<Vec<_>>();
+                write!(f, "Key({}, [{}])", self.name, kv_pairs.join(", "))
+            }
+        }
+    }
+}
+
+/// A value that can be converted into a `Key`.
+pub trait IntoKey {
```

Comment:
isnt his just `Into<Key>`?

Reply:
Yes, but we can't implement that for string types since we don't own them, and thus the custom trait.

Reply:
Why not implement `From<String> for Key` which automatically provides a `Into<Key> for String` via the [`blanket implementation`]?

[`blanket implementation`]: https://doc.rust-lang.org/std/convert/trait.Into.html#method.into

Reply:
Yeah, I think that'd work.  Apparently I forgot all about the `From` -> `Into` blanket impl when I wrote all of this... 😅

Reply:
Done!